### PR TITLE
frontend, python: allow ignoring errors that a project already exists

### LINKED ISSUE #2140: Return the existing project trying to add a project which already exists
Original issue: https://pagure.io/copr/copr/issue/2140
Opened: 2022-03-25 09:58:15
Opened by: [csomh](https://accounts.fedoraproject.org/user/csomh)

(This came up while working with the Python API, but I'm not sure if the solution is going to impact only that one.)

In Packit we hit a race condition, when trying to create the same project from tasks running in parallel. [The solution](https://github.com/packit/packit/pull/1527/files) was to check the exception and if it was raised b/c the project already exists, ignore the error. But this is ugly.

Looking at the bigger picture in that code, we also have `create_copr_project_if_not_exists()`, which first checks if the project already exists, then tries to create it. Which is 2 API calls.

The ask here is to have a new endpoint to `project_proxy.add()` to not to raise an error, if a project with an identical name and configuration (this is important, if projects are configured differently, the caller should find out about this) already exists, OR

to have a new API which either creates the project if it doesn't exist, or returns the existing project, if it does.

### PR DESCRIPTION
**Pull request by FrostyX**
Fix #2140

### CODE REVIEW

**Comment by FrostyX** [comment]
This was much more complicated than I expected.

- @nikromen Am I doing the query parameters parsing correctly? It was done with the [RequestParser](https://flask-restx.readthedocs.io/en/latest/parsing.html) but that changed in PR #2807
- I am passing the `exist_ok` value all the way into the `CoprUniqueNameValidator` and then hackily trying to access its copr instance from the outside. It works but we may want to do it differently? Maybe not using the validator and instead of `CoprsLogic.add` using `CoprsLogic.get_or_create` (we have a similar method for `CoprDirsLogic`)?

File: frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py [hunk 1]
```diff
@@ -181,28 +181,35 @@ def get(self, query, **kwargs):
         return {"items": projects, "meta": paginator.meta}
 
 
-@apiv3_projects_ns.route("/add/<ownername>")
+@apiv3_projects_ns.route("/add/<ownername>/")
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py:184]: Is this a breaking API change?  Note that some clients do not necessarily have to follow redirects.

Comment (FrostyX) [reply frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py:184]: Sorry, this is a typo. I will revert this back.

File: frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py [hunk 2]
```diff
@@ -181,28 +181,35 @@ def get(self, query, **kwargs):
...
+        form_class = forms.CoprFormFactory.create_form_cls(user=user, group=group,
+                                                           exist_ok=exist_ok)
         set_defaults(data, form_class)
         form = form_class(data, meta={"csrf": False})
 
         if not form.validate_on_submit():
+            if exist_ok:
+                for validator in form.name.validators:
+                    if isinstance(validator, forms.CoprUniqueNameValidator):
+                        return to_dict(validator.copr)
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py:212]: This deserves some in-line comment :-)

**Comment by praiskup** [comment]
otherwise lgtm

File: frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py (hunk 2 above)
Comment (FrostyX) [reply frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_projects.py]: Agreed. And thank you for poking it, there was a small bug. I fixed it, and added a comment.

**Comment by FrostyX** [comment]
Updated, and I also added a unit test.

**Review by praiskup — APPROVED** [review]
LGTM, thank you!

**Pull request merged**